zetta_utils/layer/volumetric/precomputed/precomputed.py

```diff
@@ -58,6 +59,76 @@ def _str(n: float) -> str:  # pragma: no cover
     return str(n)
 
 
+def _check_seq_is_int(seq: Sequence[int | float]) -> bool:
+    error = False
+    for k in seq:
+        error |= not float(k).is_integer()
+    return error
+
+
+def _get_ref_scale(
+    add_scales_ref: str | dict[str, Any] | None, reference_info: dict[str, Any]
+) -> dict[str, Any]:
+    # the reference scale can either be a dictionary or an existing key in `ref_info`
+    if isinstance(add_scales_ref, dict):
+        return add_scales_ref
+    # if not, search for the given key in ref_info
+    if "scales" not in reference_info:
+        raise RuntimeError("`scales` must be in `reference_info` if `add_scales_ref` is a key")
+    if add_scales_ref is None:
+        # get the highest res scale by sorting
+        add_scales_ref = _merge_and_sort_scales(reference_info["scales"], [])[0]["key"]
+    matched = list(filter(lambda x: x["key"] == add_scales_ref, reference_info["scales"]))
+    if len(matched) == 0:
+        raise RuntimeError(f'`reference_info` does not have scale "{add_scales_ref}"')
+    return matched[0]
+
+
+def _make_scale(ref: dict[str, Any], target: Sequence[int] | dict[str, Any]) -> dict[str, Any]:
+    """Make a single scale based on the reference scale"""
+    ret = {}
+    if isinstance(target, dict):
+        ret = target.copy()
+    else:
+        ret["resolution"] = target
+    multiplier = [k / v for k, v in zip(ret["resolution"], ref["resolution"])]
+
+    # fill missing values if necessary
+    if "key" not in ret:
+        ret["key"] = "_".join([str(k) for k in ret["resolution"]])
+    if "size" not in ret:
+        ret["size"] = [k / m for k, m in zip(ref["size"], multiplier)]
+    if "voxel_offset" not in ret:
+        ret["voxel_offset"] = [k / m for k, m in zip(ref["voxel_offset"], multiplier)]
+    if "chunk_sizes" not in ret:
+        ret["chunk_sizes"] = ref["chunk_sizes"]
+    if "encoding" not in ret:
+        ret["encoding"] = ref["encoding"]
+
+    # check and convert values to int
+    errored = _check_seq_is_int(ret["size"])
+    errored |= _check_seq_is_int(ret["voxel_offset"])
+    if errored:
+        raise RuntimeError(f"Computed scale {ret} does not have integer size and offsets")
+    ret["voxel_offset"] = [int(k) for k in ret["voxel_offset"]]
+    ret["size"] = [int(k) for k in ret["size"]]
+    return ret
+
+
+def _merge_and_sort_scales(
+    existing_scales: Sequence[dict[str, Any]], new_scales: Sequence[dict[str, Any]]
+) -> Sequence[dict[str, Any]]:
+    """Merge two list of scales, overwriting the old with new entries"""
+    ret_dict = dict(zip([k["key"] for k in existing_scales], existing_scales))
+    ret_dict.update(dict(zip([k["key"] for k in new_scales], new_scales)))
+    # sort scales by voxel volumes
```

Comment:
We should probably mention that the scales will get sorted in the docstring to `build` on the off chance that a user wants to do something cheeky and expects the scales to be given in the order provided by them

Reply:
Yea good call, as neuroglancer implements it the scale order does affect how it fetches the different resolutions